### PR TITLE
feat(format): add {{linksection}} token (#387)

### LINKED ISSUE #387: [FEATURE REQUEST] {{linksection}} and/or {{linkblock}} in addition to {{linkcurrent}}?
### Discussed in https://github.com/chhoumann/quickadd/discussions/380

<div type='discussions-op-text'>

<sup>Originally posted by **tauren** February  8, 2023</sup>
With `{{linkcurrent}}`, a link is created to the current file, but clicking that link opens the file scrolled to the top. This makes it difficult to maintain context, especially in larger files.

It would be great to create a link to the current section or block that the cursor is in. Perhaps adding a `{{linksection}}` or `{{linkblock}}` feature might make sense?
</div>

### PR DESCRIPTION
**Pull request by chhoumann**
Closes #387.

## What
Adds a new read-only format token **`{{linksection}}`**. Where `{{linkcurrent}}` always links to the *top* of the current file, `{{linksection}}` links to the **heading the cursor is currently under** (`[[Note#Heading]]`), so clicking the link scrolls straight to that section.

Example: a capture that files `- From {{linksection}}` into an inbox, run while your cursor is under `## Tasks` in `Source.md`, writes `- From [[Source#Tasks]]`.

## Behavior
Mirrors `{{linkcurrent}}` (same required/optional rules, case-insensitive, replace-all), and:
- **Live editor buffer** for heading detection (not the metadata cache) → a just-typed or brand-new heading resolves without waiting for reindex.
- **Duplicate headings** → disambiguated via the ancestor path (`[[Note#Parent#Heading]]`); if even that is ambiguous it falls back to a whole-file link rather than ever pointing at the *wrong* heading.
- **Sanitizes** subpath-breaking characters (`#`, `|`, `^`, `[[ ]]`).
- **Never throws** out of a capture/template run (guarded → whole-file fallback).
- Resolved in template/capture bodies and insert-after/before selectors (parity with `{{linkcurrent}}`), plus the format-syntax **preview** and **suggester** (strict-prefix regex so `{{linkc`→linkcurrent, `{{links`→linksection).

## Verification
- 2214 unit tests pass (new: pure `buildSectionSubpath`/sanitizer tests, base-replacer behavior, suggester prefix disambiguation, docs-example coverage); lint + build clean.
- **End-to-end in a real Obsidian vault**: generated links were *clicked* and confirmed to land on the correct heading line — including the two-identical-`## Notes` case routing to the right one — and a real Capture choice produced a working `[[Source#Tasks]]` link in another note.
- Reviewed via an ultracode design pass + adversarial reviewers on both the design and the diff; their blockers (recursion on a token-named heading, ambiguous-chain links, metadata-cache staleness, empty/CRLF headings) are fixed and covered by tests.

## Scope: `{{linksection}}` only
The issue is titled "linksection **and/or** linkblock". This ships `{{linksection}}`, which resolves the core request (link to the current *section* so it scrolls there). `{{linkblock}}` is intentionally **not** built: unlike every other format token it would have to **write a `^blockid` back into the note**, and that write conflicts with the capture engine's `read → format → 3-way-merge → write` pipeline in ways that can **lose the user's captured text** (the merge is word-tokenized, so a same-line capture + `^id` conflicts and aborts). A read-only variant (reuse an existing id only) would be near-useless. Decision: close #387 with this; revisit `{{linkblock}}` only if real demand appears, using the editor-transaction design (write via the editor + `view.save()` only when the source note is the open editor, never inside the API/AI/preview paths).

Docs: `docs/docs/FormatSyntax.md` (Next/unversioned) + the template-path note-relative list.

## Summary by CodeRabbit

* **New Features**
  * Added `{{LINKSECTION}}` to generate links to the nearest heading at/above the cursor (e.g., `[[Note#Heading]]`) with whole-file fallback and required/optional behavior across completion and preview/display formatting.
* **Documentation**
  * Updated format-syntax docs with `{{LINKSECTION}}` behavior, including heading selection and disambiguation.
  * Clarified PATH-SAFE Template Path handling: `{{LINKSECTION}}` stays literal; the resolved path determines the created file extension.
* **Tests**
  * Added/expanded coverage for token replacement, cursor-based heading parsing, subpath disambiguation, suggestions, and edge cases.

### CODE REVIEW

File: src/formatters/completeFormatter.ts [hunk 1]
```diff
@@ -100,6 +102,7 @@ export class CompleteFormatter extends Formatter {
 
 		output = await this.format(output);
 		output = await this.replaceLinkToCurrentFileInString(output);
+		output = this.replaceLinkToCurrentSectionInString(output);
 		output = await this.replaceCurrentFileNameInString(output);
 		output = this.replaceTargetFolderInString(output);
 		output = this.replaceTitleInString(output);
```
[the remark this answers is not in the document]
Comment (chhoumann) [reply src/formatters/completeFormatter.ts]: Fixed in 7c3da5e2. `sanitizeHeadingForSubpath` now neutralizes `{`/`}` (and the rest of Obsidian's anchor-normalizer punctuation class), so a heading literally containing a token like `## {{FILENAMECURRENT}}` becomes anchor text `FILENAMECURRENT` — no `{{...}}` remains for a later pass to rewrite. Verified e2e: cursor under `## {{FILENAMECURRENT}}` → `[[Note#FILENAMECURRENT]]` (token not rewritten).

File: src/formatters/completeFormatter.ts [hunk 2]
```diff
@@ -216,6 +221,72 @@ export class CompleteFormatter extends Formatter {
...
+		// in an editing mode (reading mode has no meaningful cursor line).
+		if (!view || view.file?.path !== file.path) return null;
+		if (view.getMode() === "preview") return null;
+
+		const editor = view.editor;
+		const cursor = editor?.getCursor();
+		if (!editor || !cursor) return null;
+
+		const headings = getMarkdownHeadings(editor.getValue().split("\n")).map(
+			(h) => ({ heading: h.text, level: h.level, line: h.line }),
```
[the remark this answers is not in the document]
Comment (chhoumann) [reply src/formatters/completeFormatter.ts]: Fixed in 7c3da5e2. Heading parsing moved to a dedicated `extractHeadingsFromLines` that skips fenced code blocks (and YAML frontmatter), bounds the level to 1–6, and allows up to 3 leading spaces. Verified e2e: a `# not a heading` line inside a ```js fence is ignored, and the cursor resolves to the nearest real heading.

**Comment by chhoumann** [comment]
Thanks for the reviews 🙏 Pushed `7c3da5e2` addressing all of it (ultracode + Codex adversarial pass on the diff, plus the inline P2s):

- **Heading-anchor sanitization** now mirrors Obsidian's internal anchor normalizer exactly (verified against `obsidian.asar`'s `LT`), replacing the full punctuation class. Fixes: stray `]]` breaking the wikilink, brackets causing wrong-heading resolution, and a heading literally containing a QuickAdd token (e.g. `## {{TITLE}}`) having that token rewritten inside the generated link.
- **Heading parse** (`extractHeadingsFromLines`): skips fenced code blocks + YAML frontmatter, bounds level to 1–6, allows ≤3 leading spaces.
- **CRLF**: split on `/\r?\n/` so a trailing `\r` no longer zeroes out heading detection.
- **Ambiguity**: a duplicate whose ancestor chain can't grow a real disambiguating segment now falls back to a whole-file link instead of a wrong-heading link.
- **Tests**: helper expectations updated to Obsidian's normalization; added `extractHeadingsFromLines` coverage and `CompleteFormatter` runtime-resolver tests (reading mode / file mismatch / no view / CRLF / throw→fallback / no-file→throw). Full suite: 2230 passing.
- **e2e** (real Obsidian vault) re-verified: CRLF, fenced-code `#`, token-named heading, bracket/`]]` heading, and the ambiguous-fallback case all behave correctly; generated links click through to the right heading line.

@codex review

File: src/formatters/helpers/sectionLink.ts [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+				fenceLen = len;
+			} else if (ch === fenceChar && len >= fenceLen) {
+				inFence = false;
+				fenceChar = "";
+				fenceLen = 0;
+			}
+			continue;
+		}
+		if (inFence) continue;
+
```
[the remark this answers is not in the document]
Comment (chhoumann) [reply src/formatters/helpers/sectionLink.ts]: Good catch — fixed in 0d6cee4a. Opening fences may carry an info string, but a closing fence must now be bare (`^\s{0,3}(`{3,}|~{3,})\s*$`), so a content line like ```ruby inside an open block no longer closes it. Added a regression test covering exactly that.

**Comment by chhoumann** [comment]
Pushed 0d6cee4a addressing CodeRabbit's fence-close finding (a closing code fence must be bare; info-string lines inside a block no longer close it). Full suite 2231 green.

@codex review

File: src/formatters/helpers/sectionLink.ts [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+			const open = line.match(/^\s{0,3}(`{3,}|~{3,})/);
+			if (open) {
+				inFence = true;
+				fenceChar = open[1][0];
+				fenceLen = open[1].length;
+				continue;
+			}
+		} else {
+			const close = line.match(/^\s{0,3}(`{3,}|~{3,})\s*$/);
+			if (
```
[the remark this answers is not in the document]
Comment (chhoumann) [reply src/formatters/helpers/sectionLink.ts:99]: Fixed in b33485a5. The heading and fence patterns now allow only ` {0,3}` (spaces) of indentation, so a leading tab is treated as an indented code line and not parsed as a heading. Added a regression test (`\t# Code` and `    # Code` → no heading).

**Comment by chhoumann** [comment]
Pushed b33485a5 for the tab-indent finding: ATX heading/fence indentation is now spaces-only (` {0,3}`), so a tab-indented `# ...` code line is no longer parsed as a heading. 2231 tests green.

@codex review

**Comment by chhoumann** [comment]
@codex review

**Comment by chhoumann** [comment]
Pushed 1c2354cc:
- **Setext headings**: `extractHeadingsFromLines` now also recognizes `Title` underlined by `===` (H1) / `---` (H2), guarded against thematic breaks (`---` after a blank), list dashes, and ATX-then-`---`. e2e-verified: cursor under a setext section links + navigates correctly.
- **app.js**: removed — that 3.6 MB Obsidian bundle was accidentally swept into an earlier commit by `git add -A` (a review step had extracted it for inspection). It's unused by build/runtime; deleted + gitignored.

Full suite 2235 green.

@codex review

**Comment by chhoumann** [comment]
Pushed 3bad57d8:
- **Case-insensitive dedup**: uniqueness + ancestor-chain collision checks now compare lower-cased keys, so `## Todo` / `## todo` are treated as duplicates (ancestor path or whole-file fallback) instead of a bare `#todo` that resolves to the first. Emitted subpath keeps original case.
- **Multi-line setext**: when the underline sits under a multi-line paragraph, we now fall back (emit no heading) rather than a partial `#title` anchor. Single-line setext unaffected.

2237 tests green.

@codex review

**Comment by chhoumann** [comment]
Pushed 5fa333c5: `isSetextContentLine` now treats `---`/`===` runs and `***`/`___` thematic breaks as structural boundaries, so a setext first heading immediately after a frontmatter close (or a thematic break) is recognized instead of skipped. Added tests. 2239 green.

@codex review

**Comment by chhoumann** [comment]
Pushed 9cd1cfca: `{{linkcurrent}}` and `{{linksection}}` now resolve in a **single non-rescanning pass** (`replaceCurrentFileLinksInString`), so neither re-scans the other's generated link — fixes the `{{linksection}}.md`-named-file corruption (and the symmetric direction), which is actually a pre-existing class for the shipped link/name tokens. Regression tests added for both directions. 2241 green.

@codex review

File: src/formatters/completeFormatter.ts [hunk 5]
```diff
@@ -99,7 +103,9 @@ export class CompleteFormatter extends Formatter {
 		let output: string = input;
 
 		output = await this.format(output);
-		output = await this.replaceLinkToCurrentFileInString(output);
+		// Resolve {{linkcurrent}} + {{linksection}} in one pass so neither
+		// re-scans the other's generated link (both embed the basename).
+		output = this.replaceCurrentFileLinksInString(output);
```
[the remark this answers is not in the document]
Comment (chhoumann) [reply src/formatters/completeFormatter.ts:108]: Good call. This is a **pre-existing** rescan class: a note literally named `{{filenamecurrent}}.md` with a `{{linkcurrent}}` body already hangs the shipped `replaceCurrentFileNameInString` while-loop, independent of this PR. This PR hardened the two **link** tokens (single non-rescanning pass); fully closing the class means folding `{{filenamecurrent}}`/`{{folder}}`/`{{title}}` into the same non-rescanning resolution, which touches shipped token behavior for a pathological input no real user produces. Tracking that hardening separately in #1358 rather than expanding this feature PR's blast radius.
